### ISSUE TITLE
JsonLogger: Malformed UTF-8 characters, possibly incorrectly encoded

### ISSUE THREAD

**Issue by niconoe-** [infection/infection#1316]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.17.3
| Test Framework version | PHPUnit 9.3.3
| PHP version | 7.4.8
| Platform    | Ubuntu 16

Hi and thank you very much for this awesome work!

When I run PHP Infection with a code that contains a base64_decode function, I have the error message "Malformed UTF-8 characters, possibly incorrectly encoded" when using the JsonLogger.

I'm not able to reproduce the issue in the playground, but I clearly identify this as when I comment my base64_decode in my project and adapt the related tests, the JSON export works like a charm.

I'm using PHP Infection with the following options:
* --coverage=/reports/phpunit (my phpunit report folder)
* --show-mutations
* --no-progress
* --threads=$(nproc)

<details>
 <summary>phpunit.xml</summary>

```xml

<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         cacheResultFile="/reports/phpunit/cache/.phpunit.result.cache"
>
    <php>
        <ini name="memory_limit" value="-1" />
    </php>
    <testsuites>
        <testsuite name="${PROJECT_NAME}">
            <directory>${TESTS_FOLDER}</directory>
        </testsuite>
    </testsuites>
    <coverage>
        <include>
            <directory suffix=".php">${SOURCES_FOLDER}</directory>
        </include>

        <report>
            <html outputDirectory="/reports/phpunit/coverage-html/"/>
            <xml outputDirectory="/reports/phpunit/coverage-xml/"/>
            <clover outputFile="/reports/phpunit/coverage-clover.xml"/>
        </report>
    </coverage>
    <logging>
        <junit outputFile="/reports/phpunit/junit.xml"/>
    </logging>
</phpunit>

```

</details>

<details>
 <summary>infection.json</summary>

```json

{
    "testFramework":"phpunit",
    "bootstrap":"vendor/autoload.php",
    "timeout": 10,
    "source": {
        "directories": [
            "${SOURCES_FOLDER}"
        ]
    },
    "phpUnit": {
        "configDir": ".",
        "customPath": "vendor/bin/phpunit"
    },
    "logs": {
        "json": "/reports/infection/infection-log.json"
    },
    "initialTestsPhpOptions": "-d memory_limit=-1 -d zend_extension=xdebug.so",
    "testFrameworkOptions": "--configuration phpunit.xml"
}

```

</details>

<details>
 <summary>Output with issue</summary>

![image](https://user-images.githubusercontent.com/9560327/91358120-8e140500-e7f2-11ea-9e59-80591c0e173c.png)

</details>

I know that sometime the `json_encode` is dealing wrong with bad character format. Is it an option to use `mb_convert_encoding` wrapping the `$data` to `json_encode`? Or do you think there's another way?

Thanks a lot.
Best regards,

**Comment by sanmai** [comment]
Can you run Infection with `-vvv` so we can get a backtrace for this error? Thanks.

**Comment by niconoe-** [comment]
Sure!
But I think the -vvv option is not working as expected:

![image](https://user-images.githubusercontent.com/9560327/91410692-921e4200-e847-11ea-9e28-5b997f6cd416.png)

Do I miss something here?
Thanks.

**Comment by sanmai** [comment]
Will it be too difficult to copy the text of the exception stack trace here as text?

**Comment by niconoe-** [comment]
No, sure. Here it is:

```
Fatal error: Uncaught Error: Call to undefined function _HumbugBox2267eabcbdcf\Symfony\Component\Console\get_debug_type() in phar:///usr/local/bin/infection/vendor/symfony/console/Application.php:484
Stack trace:
#0 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(473): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->doRenderThrowable(Object(JsonException), Object(_HumbugBox2267eabcbdcf\Symfony\Component\Console\Output\StreamOutput))
#1 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(81): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->renderThrowable(Object(JsonException), Object(_HumbugBox2267eabcbdcf\Symfony\Component\Console\Output\StreamOutput))
#2 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(101): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->_HumbugBox2267eabcbdcf\Symfony\Component\Console\{closure}(Object(JsonException))
#3 phar:///usr/local/bin/infection/bin/infection(71): _HumbugBox2267eabcbdcf\Symf in phar:///usr/local/bin/infection/vendor/symfony/console/Application.php on line 484
[debug] Restarted process exited 255
```

**Comment by sanmai** [comment]
Thank you. It seems like a different error now. Would you mind showing the output above this exception?

**Comment by niconoe-** [comment]
Then you have all :).

```
Box Requirements Checker
========================

> Using PHP 7.4.9
> PHP is using the following php.ini file:
  WARNING: No configuration file (php.ini) used by PHP!

> Checking Box requirements:
  ✔ The application requires the version "^7.3" or greater.
  ✔ The application requires the extension "zlib".
  ✔ The application requires the extension "dom".
  ✔ The application requires the extension "json".
  ✔ The application requires the extension "libxml".
  ✔ The package "nikic/php-parser" requires the extension "tokenizer".
  
                                                                                
 [OK] Your system is ready to run the application.                              
                                                                                

[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] The Xdebug extension is loaded (2.9.6)
[debug] Process restarting (INFECTION_ALLOW_XDEBUG=internal|2.9.6|1|*|*)
[debug] Running '/usr/local/bin/php' '/usr/local/bin/infection' '--coverage=/reports/phpunit' '--show-mutations' '--no-progress' '-vvv' '--threads=4' '--ansi'

Box Requirements Checker
========================

> Using PHP 7.4.9
> PHP is using the following php.ini file:
  /tmp/ilBJDJ

> Checking Box requirements:
  ✔ The application requires the version "^7.3" or greater.
  ✔ The application requires the extension "zlib".
  ✔ The application requires the extension "dom".
  ✔ The application requires the extension "json".
  ✔ The application requires the extension "libxml".
  ✔ The package "nikic/php-parser" requires the extension "tokenizer".
  
                                                                                
 [OK] Your system is ready to run the application.                              
                                                                                

[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] Restarted (240 ms). The Xdebug extension is not loaded

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version 0.17.3@d3d77a7629d237e437d5bb66606f9493bdcf369e


Running initial test suite...

PHPUnit version: 9.3.3

.: killed, M: escaped, U: uncovered, E: fatal error, T: timed out, S: skipped


Generate mutants...

Processing source code files: 0
..................................................   (   50)
............................................E.E..E   (  100)
E.EE.E.EE.E.EEEE..EEE..EEEEEEEEEEEE...............   (  150)
..................................................   (  200)
..................................................   (  250)
..................................................   (  300)
..................................................   (  350)
..................................................   (  400)
..................................................   (  450)
..................................................   (  500)
..................................................   (  550)
..........
Escaped mutants:
================


Not covered mutants:
====================



560 mutations were generated:
     531 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
      29 errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 100%
         Mutation Code Coverage: 100%
         Covered Code MSI: 100%

Please note that some mutants will inevitably be harmless (i.e. false positives).
```
Right after that, I have the exception.

**Comment by maks-rafalko** [comment]
@niconoe- unfortunately, can't reproduce it with a very simple source code and test.

Please see here https://github.com/infection/infection/compare/bugfix/base64-json-logger?expand=1 - pushed a branch with e2e test.

Could you please share what values do you use in `base64_decode`/`base64_encode` functions so that I can add them to the test.

When I run infection for this small projectm I get

<details>

<summary>infection output</summary>

```
    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version dev-master@e1beb5ce19e7fef992ed8dbaa6ff78dfbda04570


Running initial test suite...

PHPUnit version: 8.5.8

    7 [============================] < 1 sec

Generate mutants...

Processing source code files: 1/1
.: killed, M: escaped, U: uncovered, E: fatal error, T: timed out, S: skipped

..                                                   (2 / 2)

2 mutations were generated:
       2 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 100%
         Mutation Code Coverage: 100%
         Covered Code MSI: 100%

Please note that some mutants will inevitably be harmless (i.e. false positives).

Time: 0s. Memory: 14.00MB
```
</details>


and `infection.log.json` is correctly dumped to FS:

<details>
<summary>`infection.json.log`</summary>

```json
{
    "stats": {
        "totalMutantsCount": 2,
        "killedCount": 2,
        "notCoveredCount": 0,
        "escapedCount": 0,
        "errorCount": 0,
        "skippedCount": 0,
        "timeOutCount": 0,
        "msi": 100,
        "mutationCodeCoverage": 100,
        "coveredCodeMsi": 100
    },
    "escaped": [],
    "timeouted": [],
    "killed": [
        {
            "mutator": {
                "mutatorName": "PublicVisibility",
                "originalSourceCode": "<?php\n\nnamespace Base64EncodeJsonLogger;\n\nuse function base64_decode;\nuse function base64_encode;\nclass SourceClass\n{\n    public function base64encode() : string\n    {\n        return base64_encode('http:\/\/some.url\/with+spaces-');\n    }\n    public function base64decode() : string\n    {\n        return base64_decode('aHR0cDovL3NvbWUudXJsL3dpdGgrc3BhY2VzLQ==');\n    }\n}",
                "mutatedSourceCode": "<?php\n\nnamespace Base64EncodeJsonLogger;\n\nuse function base64_decode;\nuse function base64_encode;\nclass SourceClass\n{\n    protected function base64encode() : string\n    {\n        return base64_encode('http:\/\/some.url\/with+spaces-');\n    }\n    public function base64decode() : string\n    {\n        return base64_decode('aHR0cDovL3NvbWUudXJsL3dpdGgrc3BhY2VzLQ==');\n    }\n}",
                "originalFilePath": "\/Users\/maksrafalko\/sites\/infection\/tests\/e2e\/Base64EncodeJsonLogger\/src\/SourceClass.php",
                "originalStartLine": 10
            },
            "diff": "--- Original\n+++ New\n@@ @@\n use function base64_encode;\n class SourceClass\n {\n-    public function base64encode() : string\n+    protected function base64encode() : string\n     {\n         return base64_encode('http:\/\/some.url\/with+spaces-');\n     }",
            "processOutput": "PHPUnit 8.5.8 by Sebastian Bergmann and contributors.\n\nE\n\nTime: 24 ms, Memory: 4.00 MB\n\nThere was 1 error:\n\n1) Base64EncodeJsonLogger\\Test\\SourceClassTest::test_base64_encode\nError: Call to protected method Base64EncodeJsonLogger\\SourceClass::base64encode() from context 'Base64EncodeJsonLogger\\Test\\SourceClassTest'\n\n\/Users\/maksrafalko\/sites\/infection\/tests\/e2e\/Base64EncodeJsonLogger\/tests\/SourceClassTest.php:13\n\nERRORS!\nTests: 1, Assertions: 0, Errors: 1."
        },
        {
            "mutator": {
                "mutatorName": "PublicVisibility",
                "originalSourceCode": "<?php\n\nnamespace Base64EncodeJsonLogger;\n\nuse function base64_decode;\nuse function base64_encode;\nclass SourceClass\n{\n    public function base64encode() : string\n    {\n        return base64_encode('http:\/\/some.url\/with+spaces-');\n    }\n    public function base64decode() : string\n    {\n        return base64_decode('aHR0cDovL3NvbWUudXJsL3dpdGgrc3BhY2VzLQ==');\n    }\n}",
                "mutatedSourceCode": "<?php\n\nnamespace Base64EncodeJsonLogger;\n\nuse function base64_decode;\nuse function base64_encode;\nclass SourceClass\n{\n    public function base64encode() : string\n    {\n        return base64_encode('http:\/\/some.url\/with+spaces-');\n    }\n    protected function base64decode() : string\n    {\n        return base64_decode('aHR0cDovL3NvbWUudXJsL3dpdGgrc3BhY2VzLQ==');\n    }\n}",
                "originalFilePath": "\/Users\/maksrafalko\/sites\/infection\/tests\/e2e\/Base64EncodeJsonLogger\/src\/SourceClass.php",
                "originalStartLine": 15
            },
            "diff": "--- Original\n+++ New\n@@ @@\n     {\n         return base64_encode('http:\/\/some.url\/with+spaces-');\n     }\n-    public function base64decode() : string\n+    protected function base64decode() : string\n     {\n         return base64_decode('aHR0cDovL3NvbWUudXJsL3dpdGgrc3BhY2VzLQ==');\n     }\n }",
            "processOutput": "PHPUnit 8.5.8 by Sebastian Bergmann and contributors.\n\n.E                                                                  2 \/ 2 (100%)\n\nTime: 24 ms, Memory: 4.00 MB\n\nThere was 1 error:\n\n1) Base64EncodeJsonLogger\\Test\\SourceClassTest::test_base64_decode\nError: Call to protected method Base64EncodeJsonLogger\\SourceClass::base64decode() from context 'Base64EncodeJsonLogger\\Test\\SourceClassTest'\n\n\/Users\/maksrafalko\/sites\/infection\/tests\/e2e\/Base64EncodeJsonLogger\/tests\/SourceClassTest.php:19\n\nERRORS!\nTests: 2, Assertions: 1, Errors: 1."
        }
    ],
    "errored": [],
    "uncovered": []
}
```

</details>

**Comment by niconoe-** [comment]
Hi Marks, and thank you for your investigation. I wasn’t able to reproduce neither in the playground. As the incriminated snippet is from my company, I’ll try to reduce it to the strict minimum to reproduce and anonymize, then I’ll send you the sources so you could give it a try.

Thanks again and have a nice day!

**Comment by niconoe-** [comment]
Hi again Marks!

After trimming and trimming again, I finally came to a situation I am able to reproduce it using the playground!

You can take a look at https://infection-php.dev/r/o00

Here are the assets I am using on my trimmed version project:

<details>
 <summary>composer.json</summary>

```json
{
    "name": "project/infection_bug_json",
    "require": {
        "php": ">=7.4.0",
        "ext-json": "*"
    },
    "require-dev": {
        "phpunit/phpunit": "^9.3"
    },
    "autoload": {
        "psr-4": {"AcmeProject\\": "src/AcmeProject"}
    },
    "autoload-dev": {
        "psr-4": {"AcmeProject\\Tests\\": "src/Tests"}
    },
    "prefer-stable": true,
    "minimum-stability": "dev"
}
```
</details>

<details>
 <summary>infection.json</summary>

```json
{
    "testFramework":"phpunit",
    "bootstrap":"vendor/autoload.php",
    "timeout": 10,
    "source": {
        "directories": [
            "src/AcmeProject"
        ]
    },
    "phpUnit": {
        "configDir": ".",
        "customPath": "vendor/bin/phpunit"
    },
    "logs": {
        "json": "/reports/infection/infection-log.json"
    },
    "initialTestsPhpOptions": "-d memory_limit=-1 -d zend_extension=xdebug.so",
    "testFrameworkOptions": "--configuration phpunit.xml"
}

```

</details>

<details>
<summary>phpunit.xml</summary>

```xml

<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         cacheResultFile="/reports/phpunit/cache/.phpunit.result.cache"
>
    <php>
        <ini name="memory_limit" value="-1" />
    </php>
    <testsuites>
        <testsuite name="infection_bug">
            <directory>src/Tests</directory>
        </testsuite>
    </testsuites>
    <coverage>
        <include>
            <directory suffix=".php">src/AcmeProject</directory>
        </include>

        <report>
            <html outputDirectory="/reports/phpunit/coverage-html/"/>
            <xml outputDirectory="/reports/phpunit/coverage-xml/"/>
            <clover outputFile="/reports/phpunit/coverage-clover.xml"/>
        </report>
    </coverage>
    <logging>
        <junit outputFile="/reports/phpunit/junit.xml"/>
    </logging>
</phpunit>


```

</details>



Command line (as I'm using my company's docker image, you could not try it on your side, but it gives you the options in CLI I use)
<details>
<summary>Without very very verbose</summary>

```

> docker run --rm --tty --network host --volume /tmp/infection_bug/www:/app --volume /tmp/infection_bug/www:/data/www hub.aareonit.fr:5000/preprod/infection:0.17-php7.4-fpm-alpine3.12 --coverage=/reports/phpunit --no-progress --threads=$(nproc)

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version 0.17.3@d3d77a7629d237e437d5bb66606f9493bdcf369e


Running initial test suite...

PHPUnit version: 9.3.8

.: killed, M: escaped, U: uncovered, E: fatal error, T: timed out, S: skipped


Generate mutants...

Processing source code files: 0
.......

7 mutations were generated:
       7 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 100%
         Mutation Code Coverage: 100%
         Covered Code MSI: 100%

Please note that some mutants will inevitably be harmless (i.e. false positives).

In JsonLogger.php line 23:
                                                            
  Malformed UTF-8 characters, possibly incorrectly encoded  
                                                            

run [--test-framework TEST-FRAMEWORK] [--test-framework-options TEST-FRAMEWORK-OPTIONS] [-j|--threads THREADS] [--only-covered] [-s|--show-mutations] [--no-progress] [--force-progress] [-c|--configuration CONFIGURATION] [--coverage COVERAGE] [--mutators MUTATORS] [--filter FILTER] [--formatter FORMATTER] [--min-msi MIN-MSI] [--min-covered-msi MIN-COVERED-MSI] [--log-verbosity LOG-VERBOSITY] [--initial-tests-php-options INITIAL-TESTS-PHP-OPTIONS] [--skip-initial-tests] [--ignore-msi-with-no-mutations] [--debug] [--dry-run]


```


</details>

<details>
<summary>With very very verbose</summary>

```

> docker run --rm --tty --network host --volume /tmp/infection_bug/www:/app --volume /tmp/infection_bug/www:/data/www ${DOCKER_REPOSITORY}/infection:0.17-php7.4-fpm-alpine3.12 -vvv --coverage=/reports/phpunit --no-progress --threads=$(nproc)

Box Requirements Checker
========================

> Using PHP 7.4.9
> PHP is using the following php.ini file:
  WARNING: No configuration file (php.ini) used by PHP!

> Checking Box requirements:
  ✔ The application requires the version "^7.3" or greater.
  ✔ The application requires the extension "zlib".
  ✔ The application requires the extension "dom".
  ✔ The application requires the extension "json".
  ✔ The application requires the extension "libxml".
  ✔ The package "nikic/php-parser" requires the extension "tokenizer".
  
                                                                                
 [OK] Your system is ready to run the application.                              
                                                                                

[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] The Xdebug extension is loaded (2.9.6)
[debug] Process restarting (INFECTION_ALLOW_XDEBUG=internal|2.9.6|1|*|*)
[debug] Running '/usr/local/bin/php' '/usr/local/bin/infection' '-vvv' '--coverage=/reports/phpunit' '--no-progress' '--threads=4' '--ansi'

Box Requirements Checker
========================

> Using PHP 7.4.9
> PHP is using the following php.ini file:
  /tmp/CeKeFm

> Checking Box requirements:
  ✔ The application requires the version "^7.3" or greater.
  ✔ The application requires the extension "zlib".
  ✔ The application requires the extension "dom".
  ✔ The application requires the extension "json".
  ✔ The application requires the extension "libxml".
  ✔ The package "nikic/php-parser" requires the extension "tokenizer".
  
                                                                                
 [OK] Your system is ready to run the application.                              
                                                                                

[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] Restarted (144 ms). The Xdebug extension is not loaded

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version 0.17.3@d3d77a7629d237e437d5bb66606f9493bdcf369e


Running initial test suite...

PHPUnit version: 9.3.8

.: killed, M: escaped, U: uncovered, E: fatal error, T: timed out, S: skipped


Generate mutants...

Processing source code files: 0
.......

7 mutations were generated:
       7 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 100%
         Mutation Code Coverage: 100%
         Covered Code MSI: 100%

Please note that some mutants will inevitably be harmless (i.e. false positives).


Fatal error: Uncaught Error: Call to undefined function _HumbugBox2267eabcbdcf\Symfony\Component\Console\get_debug_type() in phar:///usr/local/bin/infection/vendor/symfony/console/Application.php:484
Stack trace:
#0 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(473): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->doRenderThrowable(Object(JsonException), Object(_HumbugBox2267eabcbdcf\Symfony\Component\Console\Output\StreamOutput))
#1 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(81): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->renderThrowable(Object(JsonException), Object(_HumbugBox2267eabcbdcf\Symfony\Component\Console\Output\StreamOutput))
#2 phar:///usr/local/bin/infection/vendor/symfony/console/Application.php(101): _HumbugBox2267eabcbdcf\Symfony\Component\Console\Application->_HumbugBox2267eabcbdcf\Symfony\Component\Console\{closure}(Object(JsonException))
#3 phar:///usr/local/bin/infection/bin/infection(71): _HumbugBox2267eabcbdcf\Symf in phar:///usr/local/bin/infection/vendor/symfony/console/Application.php on line 484
[debug] Restarted process exited 255

```

</details>




Do you need anything else?
Thanks a lot.
Regards,

**Comment by maks-rafalko** [comment]
Thanks @niconoe- for the reproducer. Made a PR with the fix https://github.com/infection/infection/pull/1320

**Comment by maks-rafalko** [comment]
Released as 0.17.5

**Comment by niconoe-** [comment]
Thank you very much! 👍
